pkg/logutil/merge_logger.go

```diff
@@ -0,0 +1,217 @@
+	count int        // number of lines from start field
+}
+
+func (s *status) canPrint(now time.Time) bool {
+	return s.period > 0 && s.start.Add(s.period).Before(now)
+}
+
+func (s *status) getAndReset(now time.Time) string {
+	str := fmt.Sprintf("[merged %d repeated lines in %s]", s.count, now.Sub(s.start))
+	s.start = now
+	s.count = 0
+	return str
+}
+
+// MergeLogger merges repeated log lines and prints summary log lines.
+// It doesn't merge log at DEBUG and TRACE level because they should
+// be verbose.
+//
+// For one log line, MergeLogger prints out the line when the line appears
+// at the first time. MergeLogger holds the same log line printed within
+// normalPeriod, and prints out summary log line at the end of normalPeriod.
+// It stops merging when the line doesn't appear within the
+// normalPeriod, or is flushed out.
+//
+// For cached log lines, MergeLogger always merges the repeated lines
+// until it flushes.
+type MergeLogger struct {
+	*capnslog.PackageLogger
+
+	mu      sync.Mutex // protect statusm
+	statusm map[line]*status
+}
+
+func NewMergeLogger(logger *capnslog.PackageLogger) *MergeLogger {
+	l := &MergeLogger{
+		PackageLogger: logger,
+		statusm:       make(map[line]*status),
+	}
+	go l.outputLoop()
+	return l
+}
+
+func (l *MergeLogger) Info(entries ...interface{}) {
+	l.merge(normalPeriod, line{
+		level: capnslog.INFO,
+		str:   fmt.Sprint(entries...),
+	})
+}
+
+func (l *MergeLogger) Infof(format string, args ...interface{}) {
+	l.merge(normalPeriod, line{
+		level: capnslog.INFO,
+		str:   fmt.Sprintf(format, args...),
+	})
+}
+
+func (l *MergeLogger) Notice(entries ...interface{}) {
+	l.merge(normalPeriod, line{
+		level: capnslog.NOTICE,
+		str:   fmt.Sprint(entries...),
+	})
+}
+
+func (l *MergeLogger) Noticef(format string, args ...interface{}) {
+	l.merge(normalPeriod, line{
+		level: capnslog.NOTICE,
+		str:   fmt.Sprintf(format, args...),
+	})
+}
+
+func (l *MergeLogger) Warning(entries ...interface{}) {
+	l.merge(normalPeriod, line{
+		level: capnslog.WARNING,
+		str:   fmt.Sprint(entries...),
+	})
+}
+
+func (l *MergeLogger) Warningf(format string, args ...interface{}) {
+	l.merge(normalPeriod, line{
+		level: capnslog.WARNING,
+		str:   fmt.Sprintf(format, args...),
+	})
+}
+
+func (l *MergeLogger) Error(entries ...interface{}) {
+	l.merge(normalPeriod, line{
+		level: capnslog.ERROR,
+		str:   fmt.Sprint(entries...),
+	})
+}
+
+func (l *MergeLogger) Errorf(format string, args ...interface{}) {
+	l.merge(normalPeriod, line{
+		level: capnslog.ERROR,
+		str:   fmt.Sprintf(format, args...),
+	})
+}
+
+// CachedErrorf caches the error message.
+// The cached messages are printed out only when calling Flush.
+func (l *MergeLogger) CachedErrorf(format string, args ...interface{}) {
+	l.merge(0, line{
+		level: capnslog.ERROR,
+		str:   fmt.Sprintf(format, args...),
+	})
+}
+
+// Flush flushes summary log lines, and halts merging in progress.
+func (l *MergeLogger) Flush() {
```

Comment:
It seems to be that we should have a function merging the logic in flush and the inner loop of output loop.